packages/core/src/types/utils.ts

```diff
@@ -1,14 +1,99 @@
  *
- * @see {@link https://twitter.com/mattpocockuk/status/1622730173446557697?s=20&t=NdpAcmEFXY01xkqU3KO0Mg}
- * @example
- * Prettify<{ a: string } & { b: string } & { c: number, d: bigint }>
- * => { a: string, b: string, c: number, d: bigint }
+ * Compatible with [`exactOptionalPropertyTypes`](https://www.typescriptlang.org/tsconfig#exactOptionalPropertyTypes).
+ */
+export type ExactPartial<type> = {
+  [key in keyof type]?: type[key] | undefined
+}
+
+/** Checks if {@link type} can be narrowed further than {@link type2} */
+export type IsNarrowable<type, type2> = IsUnknown<type> extends true
+  ? false
+  : undefined extends type
+  ? false
+  : IsNever<
+      (type extends type2 ? true : false) & (type2 extends type ? false : true)
+    > extends true
+  ? false
+  : true
+
+/**
+ * @internal
+ * Checks if {@link type} is `never`
+ */
+export type IsNever<type> = [type] extends [never] ? true : false
+
+/**
+ * @internal
+ * Checks if {@link type} is `unknown`
  */
-export type Prettify<T> = {
-  [K in keyof T]: T[K]
-} & {}
+export type IsUnknown<type> = unknown extends type ? true : false
+
+/** Merges two object types into new type  */
+export type Merge<obj1, obj2> = Evaluate<
+  LooseOmit<obj1, keyof obj2 extends infer key extends string ? key : never> &
+    obj2
+>
+
+/** Removes `readonly` from all properties of an object. */
+export type Mutable<type extends object> = {
+  -readonly [key in keyof type]: type[key]
+}
+
+/** Strict version of built-in Omit type */
+export type Omit<type, keys extends keyof type> = Pick<
+  type,
+  Exclude<keyof type, keys>
+>
+
+/** Makes objects destructurable. */
+export type OneOf<
+  union extends object,
+  ///
+  keys extends KeyofUnion<union> = KeyofUnion<union>,
+> = union extends infer Item
+  ? Evaluate<Item & { [K in Exclude<keys, keyof Item>]?: undefined }>
+  : never
+type KeyofUnion<type> = type extends type ? keyof type : never
+
+/** Makes {@link key} optional in {@link type} while preserving type inference. */
+// s/o trpc (https://github.com/trpc/trpc/blob/main/packages/server/src/types.ts#L6)
+export type PartialBy<type, key extends keyof type> = ExactPartial<
+  Pick<type, key>
+> &
+  Omit<type, key>
+
+///////////////////////////////////////////////////////////////////////////
+// Loose types
+
+/** Loose version of {@link Omit} */
+export type LooseOmit<type, keys extends string> = Pick<
+  type,
+  Exclude<keyof type, keys>
+>
+
+///////////////////////////////////////////////////////////////////////////
+// Union types
+
+export type UnionEvaluate<type> = type extends object ? Evaluate<type> : type
+
+export type UnionLooseOmit<type, keys extends string> = type extends any
+  ? LooseOmit<type, keys>
+  : never
+
+export type UnionOmit<type, keys extends keyof type> = type extends any
+  ? Omit<type, keys>
+  : never
+
+export type UnionPartial<type> = type extends object ? ExactPartial<type> : type
+
+export type MaybeArray<T> = T | T[]
```

Comment:
This naming introduces some confusion, because `Maybe` is associated with the algebraic type

Reply:
Which kind of algebraic type does it associate with?

Reply:
https://en.wikipedia.org/wiki/Maybe_type

https://engineering.dollarshaveclub.com/typescript-maybe-type-and-module-627506ecc5c8#950b

The correct definition of `MaybeArray<T>` would be `T[] | null`, thus `MaybeArray<T> = T | T[]` is misleading. It's essentially like saying that `Option<T> = T | T[]`

Reply:
So, `StringOrArray` or something similar would be the proper fit here

Reply:
i think it's cool. https://github.com/search?q=MaybePromise+language:TypeScript+&type=code
https://github.com/search?q=MaybeArray+language:TypeScript+&type=code

you could check repos that use similar type naming.